crates/messages/proto/messages.proto

```diff
@@ -31,137 +19,149 @@ message WorkerState {
 }
 
 message DatasetRanges {
+  option deprecated = true;
+
   string url = 1;
   repeated Range ranges = 2;
 }
 
+message BitString {
+  bytes data = 1;
+  uint64 size = 2;
+}
+
+// Worker -> Scheduler, Portal, Ping collector
 message Ping {
-  optional string worker_id = 1;
-  optional string version = 2;
+  optional string worker_id = 1 [deprecated = true];
+  repeated DatasetRanges stored_ranges = 4 [deprecated = true];
+  bytes signature = 5 [deprecated = true];
+
+  string version = 2;  // "1.0.0"
   optional uint64 stored_bytes = 3;
-  repeated DatasetRanges stored_ranges = 4;
-  bytes signature = 5;
+  string assignment_id = 6;  // "20241008T141245_242da92f7d6c"
+  // Points to the chunks in this worker's assignment list referenced by `assignment_id`.
+  BitString missing_chunks = 7;
 }
 
 message HttpHeader {
+  option deprecated = true;
+
   string name = 1;
   string value = 2;
 }
 
 message AssignedChunk {
+  option deprecated = true;
+
   string path = 1;  // "0000000000/0000808640-0000816499-b0486318"
   repeated uint32 filenames = 2;  // index in known_filenames array
 }
 
 message DatasetChunks {
+  option deprecated = true;
+
   string dataset_id = 1;  // "s3://moonbeam-evm-1"
   string download_url = 3;  // "https://moonbeam-evm-1.sqd-datasets.io/"
   repeated AssignedChunk chunks = 4;
 }
 
 message WorkerAssignment {
+  option deprecated = true;
+
   repeated DatasetChunks dataset_chunks = 1;
   repeated HttpHeader http_headers = 2;
   repeated string known_filenames = 3;  // "blocks.parquet"
 }
 
+// Scheduler -> Worker
 message Pong {
+  option deprecated = true;
+
   bytes ping_hash = 1;
   oneof status {
-    //    google.protobuf.Empty not_registered = 2;
     google.protobuf.Empty unsupported_version = 3;
     string jailed = 6;
     WorkerAssignment active = 7;
   }
 }
 
+// Portal -> Worker
 message Query {// Optional fields enforce serializing default values
-  optional string query_id = 1;
-  optional string dataset = 2;
-  optional string query = 3;
-  optional bool profiling = 4;
-  optional string client_state_json = 5;
-  bytes signature = 6;
+  string query_id = 1;
+  string dataset = 2;  // "czM6Ly9ldGhlcmV1bS1tYWlubmV0"
+  string query = 3;  // 512 KB max
   // If present, these values should be used instead of from_block and to_block in the query contents
   optional Range block_range = 7;
+  uint64 timestamp_ms = 8;
+  bytes signature = 9;  // covers all the above fields + worker's peer ID
 }
 
+// Worker -> Portal
 message QueryResult {
   string query_id = 1;
   oneof result {
-    OkResult ok = 2;
-    string bad_request = 3;
-    string server_error = 4;
-    google.protobuf.Empty no_allocation = 5;
-    google.protobuf.Empty timeout_v1 = 6 [deprecated = true];
-    string timeout = 7;
+    QueryOk ok = 2;
+    QueryError err = 8;
   }
+  optional uint32 retry_after_ms = 9;
```

Comment:
I wanted to specify it only for the `server_overloaded` and `too_many_requests` responses, but it complicates the code a lot. `QueryError` is made a separate type because it is reused across query response, worker logs and client logs. However, the logs shouldn't contain this field. But making them separate types (as was done before) really complicates the code, adds a lot of repetitions and the types start to diverge.

So I thought it's not a bad thing that the presence of this field is not dictated by the schema. On the client side it's handled separately anyway — you don't really care for what reason the server asked to delay the requests.

Reply:
It doesn't make any sense for the logs. It's a hint for the client that is only effective at the current moment and may be totally ignored. We definitely don't want it to be signed, and if it's not signed, there is no reason to include it in the logs. How do you think it can be used?